mlos_bench/mlos_bench/storage/sql_experiment.py

```diff
@@ -0,0 +1,178 @@
+        _LOG.info("Retrieve pending trials for: %s", self._experiment_id)
+        with self._engine.connect() as conn:
+            cur_trials = conn.execute(self._schema.trial.select().where(
+                column("exp_id") == self._experiment_id,
+                column("ts_end").is_(None)
+            ))
+            for trial in cur_trials.fetchall():
+                tunables = self._get_params(
+                    conn, self._schema.config_param,
+                    config_id=trial.config_id)
+                config = self._get_params(
+                    conn, self._schema.trial_param,
+                    exp_id=self._experiment_id, trial_id=trial.trial_id)
+                # Reset .is_updated flag after the assignment:
+                yield Trial(self._engine,
+                            self._tunables.copy().assign(tunables).reset(),
+                            self._experiment_id, trial.trial_id, self._opt_target, config)
+
+    def _get_config_id(self, conn, tunables: TunableGroups) -> int:
+        """
+        Get the config ID for the given tunables. If the config does not exist,
+        create a new record for it.
+        """
+        config_hash = hashlib.sha256(str(tunables).encode('utf-8')).hexdigest()
+        cur_config = conn.execute(self._schema.config.select().where(
+            column("config_hash") == config_hash
+        )).fetchone()
+        if cur_config is not None:
+            return int(cur_config.config_id)  # mypy doesn't know it's always int
+        # Config not found, create a new one:
+        config_id: int = conn.execute(self._schema.config.insert().values(
+            config_hash=config_hash)).inserted_primary_key[0]
+        self._save_params(
+            conn, self._schema.config_param,
+            {tunable.name: tunable.value for (tunable, _group) in tunables},
+            config_id=config_id)
+        return config_id
+
+    def trial(self, tunables: TunableGroups, config: Optional[Dict[str, Any]] = None):
+        _LOG.debug("Create trial: %s:%d", self._experiment_id, self._trial_id)
+        with self._engine.begin() as conn:
+            try:
+                config_id = self._get_config_id(conn, tunables)
+                conn.execute(self._schema.trial.insert().values(
+                    exp_id=self._experiment_id,
+                    trial_id=self._trial_id,
+                    config_id=config_id,
+                    ts_start=datetime.now(),
+                    status='PENDING'
+                ))
+                if config is not None:
+                    self._save_params(
+                        conn, self._schema.trial_param, config,
+                        exp_id=self._experiment_id, trial_id=self._trial_id)
+                trial = Trial(self._engine, tunables, self._experiment_id,
+                              self._trial_id, self._opt_target, config)
+                self._trial_id += 1
+                return trial
+            except Exception:
```

Comment:
```suggestion
            except SQLException:
```
?

Reply:
nope. I think we should rollback and re-throw on any exception.